error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoZipOfMonoVoidUsage.java

```diff
@@ -0,0 +1,130 @@
+    severity = ERROR,
+    tags = LIKELY_ERROR)
+public final class MonoZipOfMonoVoidUsage extends BugChecker
+    implements MethodInvocationTreeMatcher, MemberReferenceTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final String MONO = "reactor.core.publisher.Mono";
+  private static final Supplier<Type> MONO_VOID_TYPE =
+      VisitorState.memoize(generic(type(MONO), type("java.lang.Void")));
+
+  // On Mono.zip, at least one element should match empty in order to proceed.
+  private static final Matcher<ExpressionTree> MONO_ZIP_AND_WITH =
+      allOf(
+          anyMethod().onDescendantOf(MONO).namedAnyOf("zip", "zipWith"),
+          toType(MethodInvocationTree.class, hasArgumentOfType(MONO_VOID_TYPE)));
+
+  // On Mono.zip, at least one element should match empty in order to proceed.
+  private static final Matcher<ExpressionTree> STATIC_MONO_ZIP =
+      allOf(
+          staticMethod().onClass(MONO).named("zip"),
+          toType(MethodInvocationTree.class, hasArgumentOfType(MONO_VOID_TYPE)));
+
+  /** Instantiates a new {@link MonoZipOfMonoVoidUsage} instance. */
+  public MonoZipOfMonoVoidUsage() {}
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (!MONO_ZIP_AND_WITH.matches(tree, state)) {
+      return Description.NO_MATCH;
+    }
+
+    Description.Builder description = buildDescription(tree);
+
+    if (STATIC_MONO_ZIP.matches(tree, state)) {
+      ImmutableList<? extends ExpressionTree> arguments = ImmutableList.copyOf(tree.getArguments());
+
+      String replacement =
+          Streams.concat(
+                  Stream.of(
+                      arguments.stream().findFirst().map(ExpressionTree::toString).orElseThrow()),
+                  arguments.stream()
+                      .skip(1)
+                      .map(ExpressionTree::toString)
+                      .map(arg -> String.format("(%s)", arg)))
+              .collect(joining(".then"));
+      SuggestedFix staticZipUsage = SuggestedFix.replace(tree, replacement);
+      description.addFix(staticZipUsage);
+    } else {
+      SuggestedFix instanceZipUsage =
+          SuggestedFixes.renameMethodInvocation(tree, "concatWith", state);
+      description.addFix(instanceZipUsage);
+    }
+
+    return description.build();
+  }
+
+  @Override
+  public Description matchMemberReference(MemberReferenceTree tree, VisitorState state) {
+    if (!MONO_ZIP_AND_WITH.matches(tree, state)) {
+      return Description.NO_MATCH;
+    }
+
+    // Method references are expected to occur very infrequently; generating both variants of
+    // suggested fixes is not worth the trouble.
+    return describeMatch(tree);
+  }
+
+  private static Matcher<MethodInvocationTree> hasArgumentOfType(Supplier<Type> type) {
+    return (tree, state) ->
+        tree.getArguments().stream()
+            .anyMatch(arg -> ASTHelpers.isSubtype(ASTHelpers.getType(arg), type.get(state), state));
```

Comment:
This will match both `Mono<Void>` and `Mono<Integer>`, because `ASTHelpers#isSubType` uses the **type erased** type, which would be `Mono` and `Mono` in this case. 

Comparing the full types can be done using:
```suggestion
            .anyMatch(arg -> state.getTypes().isSameType(ASTHelpers.getType(arg), type.get(state)));
```

Reply:
@rickie promised to take a look 🙂

Reply:
But let me try the suggested approach.

Reply:
```java
Mono.zip(Mono.empty(), Mono.empty())
```
does not get matched, as the type of `Mono#empty()` is `Mono<java.lang.Object>`. 
I discussed it offline with @rickie and we think we can safely not flag this case. Rick said he'll take another look :pray: